orc8r/cloud/go/obsidian/reverse_proxy/reverse_proxy.go

```diff
@@ -45,13 +73,44 @@ func AddReverseProxyPaths(server *echo.Echo) (*echo.Echo, error) {
 				},
 			}
 			g := server.Group(prefix)
+			g.Use(r.activeBackendMiddleware)
 			g.Use(middleware.Proxy(middleware.NewRoundRobinBalancer(target)))
-			registeredPaths[prefix] = addr.String()
+			r.proxyBackendsByPathPrefix[prefix] = &reverseProxyBackend{
+				serverUrl: addr,
+				active:    true,
+			}
 		}
 	}
+	r.updateInactiveBackends(activePrefixes)
 	return server, nil
 }
 
+func (r *ReverseProxyHandler) activeBackendMiddleware(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		// Echo does not allow dynamically removing routes.
+		// If the backend is no longer active (i.e. the service no longer
+		// exists), return a 404.
+		path := c.Path()
+		if strings.HasSuffix(path, "/*") {
+			path = path[:len(path)-2]
+		}
+		backend, exists := r.proxyBackendsByPathPrefix[path]
+		if exists && !backend.active {
+			return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("Service not found"))
+		}
+		return next(c)
+	}
+}
+
+func (r *ReverseProxyHandler) updateInactiveBackends(activePrefixes map[string]bool) {
```

Comment:
this could use a short comment I think; explain that proxyBackendsByPathPrefix contains the union of the old proxy configuration and all the new paths that were just queried for and we do a set difference between this unioned state and the current snapshot of all path prefixes to figure out what's been deactivated